### DISCUSSION TITLE
Command criteria only for focused window

### DISCUSSION THREAD

**Discussion by ankurdhama** [i3/i3#5105] [Q&A] [answered]
Looking at the documentation it seems that the window criteria that can be specified are applied to all possible windows that matches that criteria. For ex: `[class=ABC] kill` Will kill all window that has the class ABC.

Is there any way to specify that the criteria should only be applied to currently focused window.
For ex: I want to kill the focused window only if it matches a set of criteria. Hypothetically: `[focused class=ABC] kill` , will kill currently focused window only if it has ABC class.

The workaround that I can use currently is:
`i3-msg "[id=$(xdotool getactivewindow) class=ABC] kill"`

**Comment by Airblader** [comment, marked as answer]
You have the right idea, but we do make it a bit easier for you. Matching on the i3-internal ID is better here as it'll work for windowless containers as well.

```
[con_id=__focused__ class=ABC] kill
```

---

> **con_id**
> Compares the i3-internal container ID, which you can get via the IPC interface. Handy for scripting. **Use the special value  `__focused__` to match only the currently focused window.**

-- https://i3wm.org/docs/userguide.html#list_of_commands

**Reply by ankurdhama** [reply]
I should have read the documentation more carefully. 

On another note, what I was trying to achieve was something like:
`bindsym $mod+m [con_id=__focused__ tiling] floating enable, <other commands>; [con_id=__focused__ floating] floating disable, <other commands>`;
But this causes both set of commands to execute as the first one causes the window to become floating and then the second one also matches the criteria and it also executes the commands. Is there any way to achieve this such that only one criteria matches?

**Reply by Airblader** [reply]
If "other commands" is the same in both cases, you could combine everything:

```
bindsym $mod+m [con_id=__focused__] floating toggle, <other commands>
```

**Reply by Airblader** [reply]
If the commands differ based on tiling / floating, the easiest solution might be to write an external script using the IPC to handle this instead of using the command syntax directly. I won't rule out that it could be achieved with some `mark` trickery, but a script would certainly be cleaner.

**Reply by ankurdhama** [reply]
I tried the i3-msg route as well hoping that it will exit with non zero value in case a msg doesn't get applied to any window and I can do something like
`bindsym $mod+m exec "i3-msg '[con_id=__focused__ tiling] floating enable, <A commands>' ||  i3-msg '[con_id=__focused__ floating] floating disable, <B commands>'"`
But it seems that i3-msg only exit with non zero value when there is some error.

Anyway, thanks for the reply.

**Reply by Airblader** [reply]
I meant a proper external script using the IPC directly, not i3-msg. It can fetch the focused window, check its state and take action accordingly.

In theory you can even do this without the IPC using X only, but IPC is easier.
